File: AOC_2023/Day 19/19.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>
#include <regex>

using namespace std;

using ll = long long;

struct partRatings{
    int xmas[4];
};

class Solution
{
    public:
    vector<partRatings> parts;
    map<string, vector<string>> rules;
    map<char, int> chartoIndex = {{'x', 0}, {'m', 1}, {'a', 2}, {'s', 3}};
    void init(ifstream& input){
        string line;
        regex partRatingspattern(R"(\d+)");
        regex workflowpattern(R"((\w+)\{(.*)\})");
        smatch match;
        while (getline(input, line)){
            if(line.find(":") != string::npos){
                if(regex_search(line, match, workflowpattern)){
                    string key = match[1].str();
                    string wfsstr = match[2].str();
                    vector<string> flows;
                    stringstream ss(wfsstr);
                    string workflow;
                    while (getline(ss, workflow, ','))
                        flows.push_back(workflow);
                    rules[key] = flows;
                }
            }
            else{
                partRatings pr;
                int i = 0;
                while (regex_search(line, match, partRatingspattern)){
                    pr.xmas[i] = stoi(match.str());
                    line = match.suffix();
                    i++;
                }
                if(i != 0) parts.push_back(pr);
            }
        }
    }

    bool isPartAccepted(partRatings& pr){
        string curr = "in";
        while(curr != "R" && curr != "A"){
            for(auto flow : rules[curr]){
                if(flow.find(":") != string::npos){
                    bool b = false;
                    string next = flow.substr(flow.find(":")+1), expr = flow.substr(0, flow.find(":"));
                    if(expr.find("<") != string::npos){
                        for(auto it : chartoIndex){
                            if(expr.find(it.first) != string::npos){
                                if(pr.xmas[it.second] < stoi(expr.substr(expr.find("<")+1))){
                                    curr = next;
                                    b = true;
                                    break;
                                }
                            }
                        }
                    }
                    else if(expr.find(">") != string::npos){
                        for(auto it : chartoIndex){
                            if(expr.find(it.first) != string::npos){
                                if(pr.xmas[it.second] > stoi(expr.substr(expr.find(">")+1))){
                                    curr = next;
                                    b = true;
                                    break;
                                }
                            }
                        }
                    }
                    if(b) break;
                }
                else{
                    curr = flow;
                    break;
                }
            }
        }
        return curr == "A";
    }

    void part1(){
        ll result = 0;
        for(auto it : parts){
            if(isPartAccepted(it)){
                for(int i = 0; i < 4; i++){
                    result += it.xmas[i];
                }
            }
        }
        cout << "Part 1: " << result << "\n";
    }

    ll getCombinations(vector<pair<int,int>>& xmasIntervals){
        ll result = 1;
        for(auto it : xmasIntervals){
            result *= it.second - it.first + 1;
        }
        return result;
    }

    ll helper(vector<pair<int,int>> xmasIntervals, string workflowCurr){
        ll result = 0;
        for(auto flow : rules[workflowCurr]){
            if(flow.find(":") != string::npos){
                string next = flow.substr(flow.find(":")+1), expr = flow.substr(0, flow.find(":"));
                vector<pair<int,int>> xmasIntervalsNext = xmasIntervals;
                // x < val if true => x [1, val-1] else x [val, 4000] so we update the interval accordingly
                if(expr.find("<") != string::npos){
                    for(auto it : chartoIndex){
                        if(expr.find(it.first) != string::npos){
                            int val = stoi(expr.substr(expr.find("<")+1));
                            if(val > xmasIntervalsNext[it.second].first){
                                xmasIntervalsNext[it.second].second = min(xmasIntervalsNext[it.second].second, val-1);
                                if(next == "A") result += getCombinations(xmasIntervalsNext);
                                else if(next != "R") result += helper(xmasIntervalsNext, next);
                                xmasIntervals[it.second].first = max(xmasIntervals[it.second].first, val);
                            }
                        }
                    
                    }
                }
                else if(expr.find(">") != string::npos){
                    for(auto it : chartoIndex){
                        if(expr.find(it.first) != string::npos){
                            int val = stoi(expr.substr(expr.find(">")+1));
                            if(val < xmasIntervalsNext[it.second].second){
                                xmasIntervalsNext[it.second].first = max(xmasIntervalsNext[it.second].first, val+1);
                                if(next == "A") result += getCombinations(xmasIntervalsNext);
                                else if(next != "R") result += helper(xmasIntervalsNext, next);
                                xmasIntervals[it.second].second = min(xmasIntervals[it.second].second, val);
                            }
                        }
                    }
                }
            }
            else{
                if(flow == "A") result += getCombinations(xmasIntervals);
                else if(flow != "R") result += helper(xmasIntervals, flow);
            }
        }
        return result;
    }

    void part2(){
        vector<pair<int,int>> xmasIntervals = {{1, 4000}, {1, 4000}, {1, 4000}, {1, 4000}};
        string workflowCurr = "in";
        cout << "Part 2: " << helper(xmasIntervals, workflowCurr) << "\n";
    }
};

int main(){
    ifstream in("input.txt");
    if (!in.is_open()){
        cerr << "Could not open the file - '" << "\n";
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 19/19.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

type part struct {
	xmas [4]int
}

var chartoIndex = map[rune]int{'x': 0, 'm': 1, 'a': 2, 's': 3}
var parts []part
var rules map[string][]string

func initialize(filename string) (error) {
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()
	partRatingsPattern := regexp.MustCompile(`\d+`)
	workflowPattern := regexp.MustCompile(`(\w+)\{(.*)\}`)
	scanner := bufio.NewScanner(file)
	rules = make(map[string][]string)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, ":") {
			match := workflowPattern.FindStringSubmatch(line)
			if len(match) == 3 {
				key := match[1]
				flows := strings.Split(match[2], ",")
				for i := range flows {
					flows[i] = strings.TrimSpace(flows[i])
				}
				rules[key] = flows
			}
		} else {
			var pr part
			matches := partRatingsPattern.FindAllString(line, -1)
			for i, match := range matches {
				pr.xmas[i], _ = strconv.Atoi(match)
			}
			parts = append(parts, pr)
		}
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	return nil
}

func isPartAccepted(pr part) bool {
	curr := "in"
	for curr != "R" && curr != "A" {
		for _, flow := range rules[curr] {
			if strings.Contains(flow, ":") {
				b := false
				next := flow[strings.Index(flow, ":")+1:]
				expr := flow[:strings.Index(flow, ":")]
				if strings.Contains(expr, "<") {
					for char, index := range chartoIndex {
						if strings.Contains(expr, string(char)) {
							val, _ := strconv.Atoi(expr[strings.Index(expr, "<")+1:])
							if pr.xmas[index] < val {
								curr = next
								b = true
								break
							}
						}
					}
				} else if strings.Contains(expr, ">") {
					for char, index := range chartoIndex {
						if strings.Contains(expr, string(char)) {
							val, _ := strconv.Atoi(expr[strings.Index(expr, ">")+1:])
							if pr.xmas[index] > val {
								curr = next
								b = true
								break
							}
						}
					}
				}
				if b {break}
			} else {
				curr = flow
				break
			}
		}
	}
	return curr == "A"
}

func part1() {
	result := 0
	for _, p := range parts {
		if isPartAccepted(p) {
			for _, val := range p.xmas {
				result += val
			}
		}
	}
	fmt.Println("Part 1:", result)
}

func getCombinations(xmasIntervals [][]int) int64 {
	result := int64(1)
	for _, it := range xmasIntervals {
		result *= int64(it[1] - it[0] + 1)
	}
	return result
}

func helper(xmasIntervals [][]int, curr string) int64 {
	result := int64(0)
	for _, flow := range rules[curr] {
		if strings.Contains(flow, ":") {
			next := flow[strings.Index(flow, ":")+1:]
			expr := flow[:strings.Index(flow, ":")]
			xmasIntervalsNext := make([][]int, len(xmasIntervals))
			for i, it := range xmasIntervals {
				xmasIntervalsNext[i] = make([]int, 2)
				xmasIntervalsNext[i][0] = it[0]
				xmasIntervalsNext[i][1] = it[1]
			}
			if strings.Contains(expr, "<") {
				for char, index := range chartoIndex {
					if strings.Contains(expr, string(char)) {
						val, _ := strconv.Atoi(expr[strings.Index(expr, "<")+1:])
						if val > xmasIntervalsNext[index][0] {
							xmasIntervalsNext[index][1] = min(xmasIntervalsNext[index][1], val-1)
							if next == "A" {
								result += getCombinations(xmasIntervalsNext)
							} else if next != "R" {
								result += helper(xmasIntervalsNext, next)
							}
							xmasIntervals[index][0] = max(xmasIntervals[index][0], val)
						}
					}
				}
			} else if strings.Contains(expr, ">") {
				for char, index := range chartoIndex {
					if strings.Contains(expr, string(char)) {
						val, _ := strconv.Atoi(expr[strings.Index(expr, ">")+1:])
						if val < xmasIntervalsNext[index][1] {
							xmasIntervalsNext[index][0] = max(xmasIntervalsNext[index][0], val+1)
							if next == "A" {
								result += getCombinations(xmasIntervalsNext)
							} else if next != "R" {
								result += helper(xmasIntervalsNext, next)
							}
							xmasIntervals[index][1] = min(xmasIntervals[index][1], val)
						}
					}
				}
			}
		} else{
			if flow == "A" {
				result += getCombinations(xmasIntervals)
			} else if flow != "R" {
				result += helper(xmasIntervals, flow)
			}
		}
	}
	return result
}

func part2() {
	fmt.Println("Part 2:", helper([][]int{{1, 4000}, {1, 4000}, {1, 4000}, {1, 4000}}, "in"))
}

func main(){
	filename := "input.txt"
	err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1()
	part2()
}